File: frontend/7Shoes/src/Admin/services/productService.ts
import axios from "axios";
import api from "./api";

export interface ProductForViews {
    productId: string;

    productName?: string;
    name?: string;

    basePrice?: number;
    price?: number;

    description?: string;

    productImage?: string;
    image?: string;

    brandId?: string;
    brandName?: string;

    categoryId?: string;
    categoryName?: string;

    createdDate?: string;
    createdBy?: string;

    updatedDate?: string;
    updatedBy?: string;

    isActive?: boolean;
    
    data?: any;
}

// Lấy danh sách sản phẩm
const getAll = () => api.get<{ products: ProductForViews[] }>(api.url.products).then(response => response.data.products);

// Lấy sản phẩm theo ID
const getById = (id: string) => api.get<ProductForViews>(`${api.url.products}/${id}`).then(res => res.data); 

// Thêm sản phẩm
const add = (data: ProductForViews) => api.post<ProductForViews>(api.url.products, data).then(response => response.data);

// Cập nhật sản phẩm
const update = (id: string, data: ProductForViews) => api.put<ProductForViews>(`${api.url.products}/${id}`, data).then(response => response.data);

// Xóa sản phẩm
const remove = (id: string) => api.delete<ProductForViews>(`${api.url.products}/${id}`).then(response => response.data);

//Lấy ds sán phẩm theo brand id
const getProductsByBrandId = (brandId: string) => {
    return api.get(`${api.url.products}/by-brand/${brandId}`).then(res => res.data);
};
const productService = {
    getAll,
    getById,
    add,
    update,
    remove,
    getProductsByBrandId
};

export default productService;

File: backend/Services/Services/ProductService.cs
using Domain.Entities;
using Services.Contracts.Repositories;
using Services.Contracts.Repositories.Base;
using Services.Contracts.Services;
using Services.Models.Product;

namespace Services.Services
{
    public class ProductService : IProductService
    {
        private readonly IProductRepository _productRepository;
        private readonly IUnitOfWork _unitOfWork;

        public ProductService(IProductRepository productRepository, IUnitOfWork unitOfWork)
        {
            _productRepository = productRepository;
            _unitOfWork = unitOfWork;
        }

        public async Task<(IEnumerable<ProductViewModel> products, int totalCount)>
        GetProductsAsync(int currentPage, int pageSize, bool? isActive = null)
        {
            return await _productRepository.GetProductsWithItems(currentPage, pageSize, isActive);
        }

        //public async Task<(IEnumerable<ProductViewModel> products, int totalCount)>
        //    GetProductsAsync(int currentPage, int pageSize)
        //{
        //    return await _productRepository.GetProductsWithItems(currentPage, pageSize);
        //}

        public async Task<ProductViewModel?> GetProductByIdAsync(Guid id)
        {
            return await _productRepository.GetProductById(id);
        }

        //Lấy sản phẩm theo brand id
        public async Task<IEnumerable<ProductViewModel>> GetProductsByBrandAsync(Guid? brandId)
        {
            return await _productRepository.GetProductsByBrandAsync(brandId);
        }

        public async Task<Product> CreateProductAsync(ProductCreateModel model)
        {
            try
            {
                // Tạo mới Product từ ProductCreateModel
                var product = new Product
                {
                    Id = Guid.NewGuid(),
                    Name = model.Name,
                    Price = model.Price,
                    Description = model.Description,
                    Image = model.Image,
                    BrandId = model.BrandId,
                    CategoryId = model.CategoryId,
                    CreatedDate = DateTime.UtcNow,
                    IsActive = true
                };
                await _productRepository.CreateAsync(product);
                await _unitOfWork.SaveChangesAsync();
                return product;
            }
            catch (Exception)
            {
                throw;
            }
        }

        public async Task<bool> UpdateProductAsync(ProductUpdateModel model)
        {
            if (model == null)
                throw new ArgumentNullException(nameof(model), "Product update data is required");

            // Lấy sản phẩm hiện có từ repository (ProductForViewModel)
            var existingProduct = await _productRepository.GetProductById(model.Id);
            if (existingProduct == null) return false;

            // Tạo đối tượng ProductUpdateModel để gửi vào repo update
            var productToUpdate = new ProductUpdateModel
            {
                Id = model.Id,
                Name = model.Name,
                Description = model.Description,
                Price = model.Price,
                Image = model.Image,
                CategoryId = model.CategoryId,
                BrandId = model.BrandId,
                UpdatedDate = DateTime.UtcNow,
                IsActive = model.IsActive
            };

            // Gọi repository để cập nhật sản phẩm
            await _productRepository.UpdateProduct(productToUpdate);

            // Lưu thay đổi vào database
            await _unitOfWork.SaveChangesAsync();

            // Trả về true nếu có ít nhất một bản ghi được thay đổi
            return true;
        }


        public async Task<bool> DeleteProductAsync(Guid productId)
        {
            return await _productRepository.DeleteProduct(productId);
        }
    }
}
